### PR TITLE
Port Vim patch 8.2.3018

### PR DESCRIPTION
**Pull request by yorickpeterse**
This ports Vim commit ad52f96a2d3169cb1b915c1d4a6ba26ba6e5bd0a, fixing the issue described in https://github.com/vim/vim/issues/8400. As this is my first time porting a Vim patch to NeoVim, let me know if any extra changes are needed :smiley:

### CODE REVIEW

File: src/nvim/version.c [hunk 1]
```diff
@@ -67,6 +67,7 @@ NULL
 
 // clang-format off
 static const int included_patches[] = {
+  3018,
```
Comment (seandewar) [inline src/nvim/version.c:70]: You'll want to remove this.
Updates to `included_patches` is managed automatically (and only contains patches from v8.0.x currently).

File: src/nvim/quickfix.c [hunk 2]
```diff
@@ -3891,7 +3891,15 @@ static void qf_update_buffer(qf_info_T *qi, qfline_T *old_last)
...
-      qf_winid = curwin->handle;
+      if (curwin->w_llist == qi) {
+        win = curwin;
+      } else {
+        win = qf_find_win_with_loclist(qi);
+
+        if (win == NULL) {
+          return;
+        }
+      }
```
Comment (seandewar) [inline src/nvim/quickfix.c:3902]: Patch sets `qf_winid = win->w_id` but this seems to be missing?

File: src/nvim/version.c (hunk 1 above)
Comment (YorickPeterse) [reply src/nvim/version.c]: @seandewar Ah thanks, I've removed the line.

**Comment by YorickPeterse** [comment]
@seandewar Not sure if you're the right person to ask: the builds on sr.ht pass fine for FreeBSD, but fail for OpenBSD. It seems OpenBSD produces a different format compared to other platforms:

```
Failures: 
	From test_quickfix.vim:
	Found errors in Test_qftextfunc_other_loclist():
	function RunTheTest[37]..Test_qftextfunc_other_loclist line 38: Expected ['Line 62, Col 63', 'Line 64, Col 65'] but got ['F1|62 col 63| bbb', 'F1|64 col 65| ccc']
	function RunTheTest[37]..Test_qftextfunc_other_loclist line 40: Expected ['Ligne 76, Colonne 77', 'Ligne 78, Colonne 79'] but got ['F2|76 col 77| ddd', 'F2|78 col 79| eee']
	function RunTheTest[37]..Test_qftextfunc_other_loclist line 44: Expected ['Line 32, Col 33', 'Line 34, Col 35'] but got ['F1|32 col 33| fff', 'F1|34 col 35| ggg']
	function RunTheTest[37]..Test_qftextfunc_other_loclist line 46: Expected ['Ligne 46, Colonne 47', 'Ligne 48, Colonne 49'] but got ['F2|46 col 47| hhh', 'F2|48 col 49| jjj']
```

Is this expected, or did my patch somehow break things?

**Comment by seandewar** [comment]
Looks like v8.2.1255 also needs porting.

**Comment by YorickPeterse** [comment]
> You may be able to get away with commenting out those lines and using a local function instead if @janlazo is OK with that?

Ah. Seeing how https://github.com/vim/vim/commit/d43906d2e5969288f239df851f5ad7b1dc2c7251 introduces quite a few changes, I'll change the tests to use a regular function instead of a lambda.

**Comment by seandewar** [comment]
Just tried this and it seems to fix the test:

```diff
diff --git a/src/nvim/testdir/test_quickfix.vim b/src/nvim/testdir/test_quickfix.vim
index a54baae67..9e9bcf9b6 100644
--- a/src/nvim/testdir/test_quickfix.vim
+++ b/src/nvim/testdir/test_quickfix.vim
@@ -5097,12 +5097,19 @@ func Test_qftextfunc_other_loclist()
   " window
   lexpr ['F1:10:12:one', 'F1:20:14:two']
   let w1_id = win_getid()
+  " v8.2.1255 isn't ported, so we can't use a lambdas for quickfixtextfunc yet
+  func! TextFunc1(d)
+    return map(getloclist(a:d.winid, {'id' : a:d.id,
+        \                'items' : 1}).items[a:d.start_idx-1:a:d.end_idx-1],
+        \          "'Line ' .. v:val.lnum .. ', Col ' .. v:val.col")
+  endfunc
   call setloclist(0, [], ' ',
         \ {'lines': ['F1:10:12:one', 'F1:20:14:two'],
         \  'quickfixtextfunc':
-        \    {d -> map(getloclist(d.winid, {'id' : d.id,
-        \                'items' : 1}).items[d.start_idx-1:d.end_idx-1],
-        \          "'Line ' .. v:val.lnum .. ', Col ' .. v:val.col")}})
+        \    'TextFunc1' })
+        "\    {d -> map(getloclist(d.winid, {'id' : d.id,
+        "\                'items' : 1}).items[d.start_idx-1:d.end_idx-1],
+        "\          "'Line ' .. v:val.lnum .. ', Col ' .. v:val.col")}})
   lwindow
   let w2_id = win_getid()
 
@@ -5110,12 +5117,18 @@ func Test_qftextfunc_other_loclist()
   " list window
   topleft new
   let w3_id = win_getid()
+  func! TextFunc2(d)
+    return map(getloclist(a:d.winid, {'id' : a:d.id,
+        \                'items' : 1}).items[a:d.start_idx-1:a:d.end_idx-1],
+        \          "'Ligne ' .. v:val.lnum .. ', Colonne ' .. v:val.col")
+  endfunc
   call setloclist(0, [], ' ',
         \ {'lines': ['F2:30:32:eleven', 'F2:40:34:twelve'],
         \  'quickfixtextfunc':
-        \    {d -> map(getloclist(d.winid, {'id' : d.id,
-        \                'items' : 1}).items[d.start_idx-1:d.end_idx-1],
-        \          "'Ligne ' .. v:val.lnum .. ', Colonne ' .. v:val.col")}})
+        \    'TextFunc2' })
+        "\    {d -> map(getloclist(d.winid, {'id' : d.id,
+        "\                'items' : 1}).items[d.start_idx-1:d.end_idx-1],
+        "\          "'Ligne ' .. v:val.lnum .. ', Colonne ' .. v:val.col")}})
   lwindow
   let w4_id = win_getid()

```

File: src/nvim/testdir/test_quickfix.vim [hunk 3]
```diff
@@ -5087,4 +5087,70 @@ func Test_qftextfunc()
...
 
+func Qftextfunc_other1(d)
+  return map(getloclist(d.winid, {'id' : d.id,
+    \ 'items' : 1}).items[d.start_idx-1:d.end_idx-1],
+    \ "'Line ' .. v:val.lnum .. ', Col ' .. v:val.col")
+endfunc
+
+func Qftextfunc_other2(d)
+  return map(getloclist(d.winid, {'id' : d.id,
+    \ 'items' : 1}).items[d.start_idx-1:d.end_idx-1],
```
Comment (seandewar) [inline src/nvim/testdir/test_quickfix.vim:5098]: `d` needs to be referenced as `a:d` (doesn't apply to lambdas).

File: src/nvim/testdir/test_quickfix.vim [hunk 4]
```diff
@@ -5087,4 +5087,70 @@ func Test_qftextfunc()
...
+  call setloclist(0, [], 'f')
+
+  " create a window and a location list for it and open the location list
+  " window
+  lexpr ['F1:10:12:one', 'F1:20:14:two']
+  let w1_id = win_getid()
+  call setloclist(0, [], ' ',
+        \ {'lines': ['F1:10:12:one', 'F1:20:14:two'],
+        \  'quickfixtextfunc':
+        \    Qftextfunc_other1})
```
Comment (seandewar) [inline src/nvim/testdir/test_quickfix.vim:5115]: Don't delete the original lines; comment them out instead.

Comment (seandewar) [reply src/nvim/testdir/test_quickfix.vim:5116]: Sorry, I almost missed this, but `quickfixtextfunc` takes a string value of the function's name.

```vim
        \  'quickfixtextfunc':
        \    'Qftextfunc_other2'})
```

**Comment by YorickPeterse** [comment]
@seandewar The tests are all green. Any further changes that are needed? :smiley:

**Comment by seandewar** [comment]
LGTM, we'll just need to wait on janlazo's blessing (he's _the_ patch guru!)
Thanks for your contribution! 🎉

Two style-related things that _might_ be mentioned:

- I'd maybe consider removing the extra blank lines not in the original patch ([style guide](https://neovim.io/develop/style-guide.xml#Vertical_Whitespace); generally these are only used when readibility is a concern).
- Personally, I'd move the `Qftextfunc_other` functions inside `Test_qftextfunc_other_loclist` near to where they're being used as I wouldn't expect them to be used elsewhere.

**Comment by seandewar** [comment]
v8.2.1255 was just ported in #14886, which should mean that the original test can be ported without modifications now.

**Review by seandewar — CHANGES_REQUESTED** [review]
Need to remove these two lines

**Comment by janlazo** [comment]
@YorickPeterse @seandewar Thanks.

**Pull request merged**